framework/graphics/dx12_dump_resources.h

```diff
@@ -0,0 +1,162 @@
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+** FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef GFXRECON_DECODE_DX12_DUMP_RESOURCES_H
+#define GFXRECON_DECODE_DX12_DUMP_RESOURCES_H
+
+#include "decode/api_decoder.h"
+#include "decode/handle_pointer_decoder.h"
+#include "decode/dx_replay_options.h"
+#include "decode/dx12_object_info.h"
+#include "decode/dx12_browse_consumer.h"
+#include "graphics/dx12_util.h"
+#include "util/defines.h"
+#include "util/json_util.h"
+
+#include <d3d12.h>
+
+#include <memory>
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(graphics)
+
+const gfxrecon::util::JsonFormat kDefaultDumpResourcesFileFormat = gfxrecon::util::JsonFormat::JSON;
+
+struct Dx12DumpResourcesConfig
+{
+    std::string               captured_file_name;
+    decode::DumpResourcesType type{ decode::DumpResourcesType::kNone };
+    int                       argument{ 0 };
+};
+
+struct CopyResourceData
+{
+    format::HandleId           source_resource_id{ format::kNullHandleId };
+    uint64_t                   size{ 0 };
+    D3D12_RESOURCE_DESC        desc{};
+    dx12::ID3D12ResourceComPtr before_resource{ nullptr }; // copy resource before drawcall
+    dx12::ID3D12ResourceComPtr after_resource{ nullptr };  // copy resource after drawcall
+};
+
+struct DescriptorHeapData
+{
+    std::vector<CopyResourceData> copy_constant_buffer_resources;
+    std::vector<CopyResourceData> copy_shader_resources;
+};
+
+struct TrackDumpResources
+{
+    decode::TrackDumpCommandList target{};
+
+    // vertex
+    std::vector<CopyResourceData> copy_vertex_resources;
+
+    // index
+    CopyResourceData copy_index_resource;
+
+    // descriptor
+    std::vector<DescriptorHeapData> descriptor_heap_datas;
+
+    // render target
+    std::vector<format::HandleId>            render_target_heap_ids;
+    std::vector<D3D12_CPU_DESCRIPTOR_HANDLE> replay_render_target_handles;
+    std::vector<CopyResourceData>            copy_render_target_resources;
+    format::HandleId                         depth_stencil_heap_id{ format::kNullHandleId };
+    D3D12_CPU_DESCRIPTOR_HANDLE              replay_depth_stencil_handle{ decode::kNullCpuAddress };
+    CopyResourceData                         copy_depth_stencil_resource;
+
+    // record BeginRenderPass parameters
+    std::vector<D3D12_RENDER_PASS_ENDING_ACCESS> record_render_target_ending_accesses;
+    D3D12_RENDER_PASS_ENDING_ACCESS              record_depth_ending_access{};
+    D3D12_RENDER_PASS_ENDING_ACCESS              record_stencil_ending_access{};
+    D3D12_RENDER_PASS_FLAGS                      record_render_pass_flags{ D3D12_RENDER_PASS_FLAG_NONE };
+
+    ~TrackDumpResources() {}
+};
+
+// TODO: This class copys a lot of code to write json from VulkanExportJsonConsumerBase.
+//       We might need a class for writing json.
+class Dx12DumpResources
```

Comment:
`Dx12DumpResources` output json and resources. A lot of json code are copied from `VulkanExportJsonConsumerBase`. We should have a common class for them.